PROTOCOL.md

```diff
@@ -995,6 +1065,178 @@ Within the checkpoint, the `remove` struct does not contain the `stats` and `tag
+separate [sidecar files](#sidecar-files). These sidecar files will then be referenced in the V2 spec checkpoint.
+The sidecar files reside in the `_delta_log/_sidecars` directory.
+- A V2 spec Checkpoint could reference zero or more [sidecar file actions](#sidecar-file-information).
+
+Note: A V2 spec Checkpoint can either have all the [add and remove file](#Add-File-and-Remove-File) actions
+embedded inside itself or all of them should be in [sidecar files](#sidecar-files). Having partial
+add and remove file actions in V2 Checkpoint and partial entries in sidecar files is not allowed.
+
+After producing a V2 spec checkpoint, a writer can choose to embed some or all of the V2 spec checkpoint in
+the `_last_checkpoint` file, so that readers don't have to read the V2 Checkpoint.
+
+E.g. showing the content of V2 spec checkpoint:
+```
+{"checkpointMetadata":{"flavor":"flat","tags":{}}}
+{"metaData":{...}}
+{"protocol":{...}}
+{"txn":{"appId":"3ba13872-2d47-4e17-86a0-21afd2a22395","version":364475}}
+{"txn":{"appId":"3ae45b72-24e1-865a-a211-34987ae02f2a","version":4389}}
+{"sidecarFile":{"path":"3a0d65cd-4056-49b8-937b-95f9e3ee90e5.parquet","sizeInBytes":2341330,"modificationTime":1512909768000,"type":"fileaction","tags":{}}
+{"sidecarFile":{"path":"016ae953-37a9-438e-8683-9a9a4a79a395.parquet","sizeInBytes":8468120,"modificationTime":1512909848000,"type":"fileaction","tags":{}}
+```
+
+Another example of a v2 spec checkpoint without sidecars:
+```
+{"checkpointMetadata":{"flavor":"flat","tags":{}}}
+{"metaData":{...}}
+{"protocol":{...}}
+{"txn":{"appId":"3ba13872-2d47-4e17-86a0-21afd2a22395","version":364475}}
+{"add":{"path":"date=2017-12-10/part-000...c000.gz.parquet",...}
+{"add":{"path":"date=2017-12-09/part-000...c000.gz.parquet",...}
+{"remove":{"path":"date=2017-12-08/part-000...c000.gz.parquet",...}
+```
+
+#### V1 Spec
+
+V1 Spec do not support [sidecar files](#sidecar-files) and [checkpoint metadata](#checkpoint-metadata).
+These are flat checkpoints which contains all actions mentioned [here](#checkpoints-1).
+
+### Checkpoint Naming Scheme
+Delta supports following three checkpoint naming scheme:
+
+#### UUID-named checkpoint
+This naming scheme represents a [V2 spec checkpoint](#v2-spec) with following file name: `n.checkpoint.u.{json/parquet}`,
+where `u` is a UUID and `n` is the snapshot version that this checkpoint represents.
+The UUID-named checkpoints may be in JSON or parquet format. Since these are following [V2 spec](#v2-spec), so they must
+have a [checkpoint metadata](#checkpoint-metadata) action and may reference zero or more checkpoint [sidecar files](#sidecar-files).
+
+Example-1: Json UUID-named checkpoint with sidecars
+
+```
+00000000000000000010.checkpoint.80a083e8-7026-4e79-81be-64bd76c43a11.json
+_sidecars/3a0d65cd-4056-49b8-937b-95f9e3ee90e5.parquet
+_sidecars/016ae953-37a9-438e-8683-9a9a4a79a395.parquet
+_sidecars/7d17ac10-5cc3-401b-bd1a-9c82dd2ea032.parquet
```

Comment:
The sidecars need to appear in lexical order.

Somewhere in this section, should we make clear that the `_sidecars` dir can contain any number of other files as well?
```
00000000000000000010.checkpoint.80a083e8-7026-4e79-81be-64bd76c43a11.json
_sidecars/016ae953-37a9-438e-8683-9a9a4a79a395.parquet
  ...
_sidecars/3a0d65cd-4056-49b8-937b-95f9e3ee90e5.parquet
  ...
_sidecars/7d17ac10-5cc3-401b-bd1a-9c82dd2ea032.parquet
  ...
```
Or is that implied/obvious?

Reply:
The sidecar directory is a common dir to store sidecar files for all the checkpoints. It seems implied?